ledger-api/grpc-definitions/com/daml/ledger/api/v1/completion.proto

```diff
@@ -32,6 +32,37 @@ message Completion {
   // Optional
   string transaction_id = 3;
 
+  // The application ID that was used for the submission, as described in ``commands.proto``.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Optional
+  // TODO: Can be made required once all Daml ledgers return the application ID
+  string application_id = 6;
+
+  // The set of parties on whose behalf the commands were executed.
+  // Contains the union of `party` and `act_as` from ``commands.proto``.
+  // The order of the parties need not be the same as in the submission.
+  // Each element must be a valid PartyIdString (as described in ``value.proto``).
+  // Optional
+  // TODO: Can be made required once all Daml ledgers return the application ID
+  repeated string act_as = 7;
+
+  // The submission ID this completion refers to, as described in ``commands.proto``.
+  // Optional
+  // TODO: Can be made required once all Daml ledgers return the application ID
+  // TODO: Do we want to enforce the format of a UUID here?
+  string submission_id = 4;
+
+  // The submission rank of the submission this completion refers to, as described in ``commands.proto``.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  // Optional
+  // TODO: Can be made required once all Daml ledgers return the application ID
+  string submission_rank = 5;
+
+  // Whether this completion is a definite answer for the deduplication and rank guarantees,
+  // as defined in ``commands.proto``.
+  // Completions with status code `OK` are always definite.
+  bool definite_answer = 8;
```

Comment:
This field alone is not enough to enable the application to react correctly to non-successful commands. Some rejections are final (e.g., a command that is not well authorised), some are retryable (e.g., contention on contract key), some are retryable in slightly different ways (e.g., ledger time model failure). Wouldn't it be better to standardise the error codes we use, or come up with our own error type?

Reply:
I agree that this information is not enough to determine an appropriate reaction. I do not intend to solve the error code problem as part of the command deduplication discussion though. I therefore propose to add this field to decouple the discussions. Even if we standardize on some error code scheme, whether something is a definite answer or not is IMO cross-cutting through the error hierarchy. I don't think it's a good idea to structure the hierarchy along the definite answer property.

Regarding error codes: IMO plain gRPC error codes are not enough to convey enough information about appropriate reactions of the applications. There are competing proposals circulating at the moment, like [gRPC's richer error model](https://cloud.google.com/apis/design/errors#error_model) and rolling our own system of error codes (https://github.com/DACH-NY/canton/issues/5853 gives an entry point into the Canton discussion).

Reply:
I agree that we don't want to solve the full error problem here.

However, I would expect this PR to yield a protocol that enables applications to properly handle duplicates. AFAIR, an application requires some information (at least the `completion_offset`) of the completion that caused a `DUPLICATE` error. How do you propose handling this?

Reply:
There's now a write-up of how to communicate error details in `error.proto`.